backend/database.js

```diff
@@ -52,57 +52,91 @@ export default ((sbp('sbp/selectors/register', {
     let counter = 0
-    let { hash: currentHash, ...serverMeta } = (
-      await sbp('chelonia/db/getEntryMeta', contractID, height)
-    ) || {}
+    let currentHeight = height
+    let currentHash, serverMeta
     let prefix = ''
-    let ended = false
-    // NOTE: if this ever stops working you can also try Readable.from():
-    // https://nodejs.org/api/stream.html#stream_stream_readable_from_iterable_options
-    const stream = new Readable({
-      construct (cb) {
-        this.push('[')
-        cb()
-      },
-      read (): void {
-        if (ended) {
-          this.destroy()
-          return
+    const nextKeyOp = (() => {
+      let index
+
+      return async () => {
+        if (!index) {
+          index = (await sbp('chelonia.db/get', `_private_keyop_idx_${contractID}_${currentHeight - currentHeight % 10000}`))?.split('\x00')
         }
-        if (currentHash && counter < limit) {
-          sbp('chelonia/db/getEntry', currentHash).then(async entry => {
-            if (entry) {
-              const currentPrefix = prefix
-              prefix = ','
-              counter++
-              const { hash: newCurrentHash, ...newServerMeta } = (
-                await sbp('chelonia/db/getEntryMeta', contractID, entry.height() + 1)
-              ) || {}
-              this.push(
-                `${currentPrefix}"${strToB64(
-                  JSON.stringify({ serverMeta, message: entry.serialize() })
-                )}"`
-              )
-              currentHash = newCurrentHash
-              serverMeta = newServerMeta
-            } else {
-              this.push(']')
-              this.push(null)
-              ended = true
-            }
-          }).catch(e => {
-            console.error(`[backend] streamEntriesAfter: read(): ${e.message}:`, e.stack)
-            this.push(']')
-            this.push(null)
-            ended = true
-          })
+        const value = index?.find((h, i) => {
+          if (Number(h) >= currentHeight) {
+            // Remove values that no longer are relevant from the index
+            index = ((index: any): any[]).splice(i + 1)
+            return true
+          } else {
+            return false
+          }
+        })
+        if (value != null) {
+          const newHeight = Number(value)
+          currentHeight = newHeight
         } else {
-          this.push(']')
-          this.push(null)
-          ended = true
+          currentHeight = currentHeight - currentHeight % 10000 + 10000
+          index = undefined
+          if (currentHeight > latestHEADinfo.height) {
+            return false
+          } else {
+            return null
+          }
         }
+        return true
       }
-    })
+    })()
+    const fetchMeta = async () => {
+      if (currentHeight > latestHEADinfo.height) {
+        return false
+      }
+      const meta = await sbp('chelonia/db/getEntryMeta', contractID, currentHeight)
+      if (!meta) {
+        return false
+      }
```

Comment:
Hmm... just wondering, would this be an unusual situation, perhaps deserving of a `console.warn`?

Reply:
Not necessarily an unusual situation, it can just be the end of the contract, although I've added height guards now. It could also happen if the contract is being simultaneously deleted (granted, a strange situation, but then also nothing actionable to do)